In [26]:
%matplotlib qt
from matplotlib import pyplot as plt
from numpy.random import rand
from skimage.io import imread, imsave
from skimage.util import img_as_float
from skimage.transform import rotate
from skimage.filters import gaussian as ski_gaussian
from pylab import ginput
from scipy.signal import convolve, convolve2d, correlate2d
from scipy.signal import  gaussian as scipy_gaussian
import numpy as np
import os

os.chdir("D:\KU\Masters\Signal and image processing\Git code\SIP\Week1\Mats")

Example 1.1

In [5]:
A = imread('bigben.png')
print(type(A))
print(A.shape)
print(A.dtype)
imsave('testimage.jpg',A)

<class 'numpy.ndarray'>
(480, 640, 3)
uint8


Example 1.2

In [6]:
A = imread('cameraman.tif')
plt.figure(1)
plt.imshow(A)

plt.figure(2)
plt.axis('off')
plt.imshow(A,cmap='gray')
plt.show()

Example 1.3

In [7]:
B = rand(256,256)*125
plt.figure(1)
plt.axis('off')
plt.imshow(B,cmap='gray')
plt.colorbar()

plt.figure(2)
plt.axis('off')
plt.imshow(B,vmin=0,vmax=255,cmap='gray')
plt.colorbar()
plt.show()

Example 1.4

In [8]:
img = 'cell.tif'
img2 = 'DeltaE_16bit_gamma2.2.tif'
img3 = 'onion.png'

# img = "mat1.jpg"
# img2 = "mat2.jpg"
# img3 = "mat3.jpg"

# img = "1h_B1.tif"
# img2 = "1h_B2.tif"
# img3 = "mat3.jpg"


B = imread(img)
print(B.shape,B.dtype)
C = imread(img2)
print(C.shape,C.dtype)
D = imread(img3)
print(D.shape,D.dtype)

plt.subplot(3,1,1)
plt.imshow(B,cmap='jet')
plt.axis('off')

plt.subplot(3,1,2)
plt.imshow(C, cmap='gray')
plt.axis('off')
plt.subplot(3,1,3)
plt.imshow(D)
plt.show()

C:\Users\Birk Dissing\AppData\Local\Temp\ipykernel_11220\105191895.py:21: MatplotlibDeprecationWarning: Auto-removal of overlapping axes is deprecated since 3.6 and will be removed two minor releases later; explicitly call ax.remove() as needed.
  plt.subplot(3,1,1)
Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).


(159, 191) uint8
(2048, 3072, 3) uint16
(135, 198, 3) uint8


Example 1.5

In [9]:
plt.imshow(img_as_float(C),cmap='gray')
plt.axis('off')
plt.show()

Example 1.6

In [10]:
print(B[500:600][45:55])

[]


In [11]:
B = imread(img)

plt.figure(1)
plt.imshow(B, cmap='gray')

plt.figure(2)
plt.imshow(D)

print('Click on 3 points in the image')
coord = ginput(3)
print('You clicked: ' + str(coord))
# print pixel value at location (25,50)
print(B[25,50])
# set pixel value at (25,50) to white
B[25,50]= 255;
# view resulting changes in image
plt.figure(3)
plt.imshow(B, cmap='gray')
# print RGB pixel value at location (25,50)
print(D[25,50,:])
# print only the red value at (25,50)
print(D[25,50,0])
D[25,50,:] = [255, 255, 255]; # set pixel value to RGB white
plt.figure(4)
plt.imshow(D)
plt.show()

Click on 3 points in the image


KeyboardInterrupt: 

Example 1.7

In [ ]:
D = imread(img3)
D_read_as_gray = imread(img3, as_gray=True)
D_naive_gray = 1/3*+(D[:,:,0]+D[:,:,1]+D[:,:,2])
D_better_gray = 0.2989*D[:,:,0] + 0.5870*D[:,:,1] + 0.1140*D[:,:,2]
plt.subplot(2,2,1)
plt.title("Original")
plt.imshow(D)

plt.subplot(2,2,2)
plt.title("Read as grey")
plt.imshow(D_read_as_gray)

plt.subplot(2,2,3)
plt.title("Naive gray")
plt.imshow(D_naive_gray)

plt.subplot(2,2,4)
plt.title("Bettergrey")
plt.imshow(D_better_gray)

plt.show()

Example 1.8

In [ ]:
D = imread(img3)
Dred = D[:,:,0]
Dgreen = D[:,:,1]
Dblue = D[:,:,2]

plt.subplot(2,2,1)
plt.imshow(D)

plt.subplot(2,2,2)
plt.imshow(Dred)
plt.title("Red")

plt.subplot(2,2,3)
plt.imshow(Dgreen)
plt.title("Green")

plt.subplot(2,2,4)
plt.imshow(Dblue)
plt.title("Blue")

plt.show()


Example 1.9

In [ ]:
I = imread("cameraman.tif")
plt.imshow(I, cmap="gray")
x = [100, 100, 230]
y = [100, 200, 200]

plt.plot(x, y, 'r*')

plt.plot(x[::2], y[::2], "g-")
plt.show()
print(y[::2])

[100, 200]


Exercise 1.1

Exercise 1.2

Exercise 1.3

Exercise 1.4

Example 2.1

In [13]:
f = np.ones(64)
f = f/sum(f)

g = convolve(f,f)
g = g / sum(g)

h = convolve(g,g)
h = h / sum(g)

j = convolve(h,h)
j = j / sum(j)


plt.subplot(2,2,1)
plt.plot(f,'k-')
plt.axis('off')
plt.subplot(2,2,2)
plt.plot(g,'k-')
plt.axis('off')
plt.subplot(2,2,3)
plt.plot(h,'k-')
plt.axis('off')
plt.subplot(2,2,4)
plt.plot(j,'k-')
plt.axis('off')
plt.show()


Example 2.2

In [17]:
A = img_as_float(imread("trui.png"))
gaussian1d = scipy_gaussian(5, 2)
PSF = np.outer(gaussian1d, gaussian1d)
motionWidth = 21
h = np.zeros((motionWidth, motionWidth))
h[motionWidth//2, :] = 1
h = rotate(h, 45).astype(float)
h = h /np.sum(h)

B = convolve2d(PSF, A)
C = convolve2d(A, h)
D = convolve2d(A, A)
E = correlate2d(A, A)

plt.subplot(2, 3, 1)
plt.imshow(A, cmap="gray")

plt.subplot(2, 3, 2)
plt.imshow(B, cmap="gray")

plt.subplot(2, 3, 3)
plt.imshow(C, cmap="gray")

plt.subplot(2, 3, 4)
plt.imshow(D, cmap="gray")

plt.subplot(2, 3, 5)
plt.imshow(E, cmap="gray")

plt.show()

Example 2.3

In [30]:
A = imread("cameraman.tif")
rows, cols = A.shape

Abuild = np.zeros(A.shape)
sub = np.random.rand(rows, cols) < 0.01
Abuild[sub] = A[sub]
B10 = ski_gaussian(Abuild, sigma=2)
plt.subplot(1,2,1)
plt.axis("off")
plt.imshow(Abuild, cmap="gray")
plt.colorbar()
plt.title("Object points")

plt.subplot(1,2,2)
plt.axis("off")
plt.imshow(B10, cmap="gray")
plt.colorbar()
plt.title("Response of LSI system")
plt.show()


Exercise 2.1

Example 3.1

Example 3.2

Example 3.3

Example 3.4

Example 3.5

Example 3.6

Example 3.7

Example 3.8

Example 3.9

Example 3.10

Example 3.11